src/bounding_box.rs

```diff
@@ -0,0 +1,593 @@
+use arrayvec::ArrayVec;
+use nalgebra::{Point3, Scalar, Vector3};
+use num_traits::{Bounded, Zero};
+
+use crate::convert::clamp_cast_i32_to_u32;
+
+/// World-origin-based axis-aligned bounding box contains the entire given
+/// geometry and defines an envelope aligned to the world (euclidean) coordinate
+/// system.
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct BoundingBox<T>
+where
+    T: Scalar,
+{
+    minimum_point: Point3<T>,
+    maximum_point: Point3<T>,
+}
+
+impl<T: Bounded + Scalar + Zero + PartialOrd> BoundingBox<T> {
+    /// Creates a new bounding box. The two input points will be deconstructed
+    /// and a new couple of points will be created: minimum point with minimum
+    /// values of x, y, z and maximum point with maximum values of x, y, z. The
+    /// resulting bounding box will be defined in the units of the input points.
+    pub fn new(box_corner1: &Point3<T>, box_corner2: &Point3<T>) -> Self {
+        BoundingBox {
+            minimum_point: Point3::new(
+                if box_corner1.x < box_corner2.x {
+                    box_corner1.x
+                } else {
+                    box_corner2.x
+                },
+                if box_corner1.y < box_corner2.y {
+                    box_corner1.y
+                } else {
+                    box_corner2.y
+                },
+                if box_corner1.z < box_corner2.z {
+                    box_corner1.z
+                } else {
+                    box_corner2.z
+                },
+            ),
+            maximum_point: Point3::new(
+                if box_corner1.x > box_corner2.x {
+                    box_corner1.x
+                } else {
+                    box_corner2.x
+                },
+                if box_corner1.y > box_corner2.y {
+                    box_corner1.y
+                } else {
+                    box_corner2.y
+                },
+                if box_corner1.z > box_corner2.z {
+                    box_corner1.z
+                } else {
+                    box_corner2.z
+                },
+            ),
+        }
+    }
+
+    /// Creates a new bounding box from an iterator of points. The resulting
+    /// bounding box will encompass all the input points. The resulting bounding
+    /// box will be defined in the units of the input points. If the input list
+    /// is empty, a zero size box at the word origin will be created.
+    pub fn from_points<'a, I>(points: I) -> Option<Self>
+    where
+        I: IntoIterator<Item = &'a Point3<T>> + Clone,
+    {
+        // Return None if there are no input points.
+        points.clone().into_iter().peekable().peek()?;
```

Comment:
`points.clone().into_iter().next()?;`

Reply:
Actually neither :frowning_face: . I just realized we are cloning collections left and right. Stay tuned. There has to be a better way.

Reply:
Fortunately, for your case (at least here), there is a solution:

1) Remove the `Clone` bound
2) Use:
```rust
    let mut points_iter = points.into_iter().peekable();
    points_iter.peek()?;
    // ... later use points_iter
```

I'll add a trello task to hunt down the clones all over the project. I know you have the trick from me. I just didn't realize we are not cloning the iterator, but the original source.

Reply:
https://github.com/subdgtl/HURBAN-Selector/pull/117

Reply:
Anyway, the guideline is that you want to use `IntoIterator` over `Iterator` as a bound, because it is more generic. But if you need to clone the iterator, you need `Iterator + Clone`, and shouldn't use `IntoIterator + Clone`, because that could accidentally clone the collection. We had a couple of these bounds incorrect in the codebase, but fortunately we were not cloning collections, just iterators.